core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -3523,15 +3523,22 @@ object Stream extends StreamLowPriority {
     * All elements that are available, up to the specified limit,
     * are dequeued and emitted as a single chunk.
     */
-  def fromQueueUnterminated[F[_]: Functor, A](
+  def fromQueueUnterminated[F[_], A](
       queue: QueueSource[F, A],
       limit: Int = Int.MaxValue
-  ): Stream[F, A] =
-    fromQueueNoneTerminatedSingletons_[F, A](
-      queue.take.map(a => Some(a)),
-      queue.tryTake.map(_.map(a => Some(a))),
-      limit
-    )
+  )(implicit F: Functor[F]): Stream[F, A] = {
+    F match {
+      case f0: Async[F] =>
+        Stream.evalSeq(queue.tryTakeN(None)(f0)).repeat
```

Comment:
Isn't this busy-polling? Or, I miss the part where it blocks if the queue is empty.

Reply:
I guess the `takeN` adds the semantics of "if queue is empty block the taker fiber until there are any elements. 

Not sure what you mean by "no chance of success". The problem is, this method is public and therefore probably used by applications, which may be relying on the blocking semantics of this method.

However, it is still possible to add this or other elements, _as a different method_.

Reply:
Busy polling is now hopefully removed.